extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/tracing/TracerProcessor.java

```diff
@@ -0,0 +1,125 @@
+
+    @BuildStep(onlyIf = TracerEnabled.class)
+    CapabilityBuildItem tracingCapability() {
+        return new CapabilityBuildItem(Capability.OPENTELEMETRY_TRACER);
+    }
+
+    @BuildStep(onlyIf = TracerEnabled.class)
+    UnremovableBeanBuildItem ensureProducersAreRetained(
+            CombinedIndexBuildItem indexBuildItem,
+            Capabilities capabilities,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+
+        if (capabilities.isPresent(Capability.OPENTRACING) ||
+                capabilities.isPresent(Capability.SMALLRYE_OPENTRACING)) {
+            throw new ConfigurationException("The OpenTelemetry extension tracer can not be used in " +
+                    "conjunction with either the SmallRye OpenTracing or Jaeger extensions.");
+        }
+
+        additionalBeans.produce(AdditionalBeanBuildItem.builder()
+                .setUnremovable()
+                .addBeanClass(TracerProducer.class)
+                .build());
+
+        IndexView index = indexBuildItem.getIndex();
+
+        // Find all known SpanExporters and SpanProcessors
+        Collection<String> knownClasses = new HashSet<>();
+        knownClasses.add(SPAN_EXPORTER.toString());
+        index.getAllKnownImplementors(SPAN_EXPORTER)
+                .forEach(classInfo -> knownClasses.add(classInfo.name().toString()));
+        knownClasses.add(SPAN_PROCESSOR.toString());
+        index.getAllKnownImplementors(SPAN_PROCESSOR)
+                .forEach(classInfo -> knownClasses.add(classInfo.name().toString()));
+
+        Set<String> retainProducers = new HashSet<>();
+
+        for (AnnotationInstance annotation : index.getAnnotations(DotNames.PRODUCES)) {
+            AnnotationTarget target = annotation.target();
+            switch (target.kind()) {
+                case METHOD:
+                    MethodInfo method = target.asMethod();
+                    String returnType = method.returnType().name().toString();
+                    if (knownClasses.contains(returnType)) {
+                        retainProducers.add(method.declaringClass().name().toString());
+                    }
+                    break;
+                case FIELD:
+                    FieldInfo field = target.asField();
+                    String fieldType = field.type().name().toString();
+                    if (knownClasses.contains(fieldType)) {
+                        retainProducers.add(field.declaringClass().name().toString());
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        return new UnremovableBeanBuildItem(new UnremovableBeanBuildItem.BeanClassNamesExclusion(retainProducers));
+    }
+
+    //TODO This needs modification to use conditional behaviors when it's present
+    @BuildStep(onlyIf = TracerEnabled.class)
+    FilterBuildItem addVertxTracerFilter() {
+        return new FilterBuildItem(new VertxTracerFilter(), Integer.MAX_VALUE);
+    }
```

Comment:
You can check the prensence if a class to enabled the filter only if vertx-http is present (this is linked to the vertx-http library being optional)

Reply:
Unfortunately that isn't sufficient.

The presence of the build step still requires it be compilable, which fails if the dependency isn't present. I tested various combinations and nothing worked

Reply:
> The presence of the build step still requires it be compilable

Just saw it, it was a flaw in the conception of the `FilterBuildItem`, all those kind of build item usually provides a constructor via a String to avoid this kind of issue and be more flexible
But I'm surprised that if you return null at build time you really have an issue ...

Reply:
It won't even compile the build step because of a CNFE on `FilterBuildItem`

Reply:
Oh right, `FilterBuildItem` should be moved to a SPI for this to work.
This is the design flaw ;)